Minimize test images customization

ansasaki opened the pull request:
Minimize test images customization to make possible to easily test other images.

Koncpa commented:
@ansasaki Just FYI for running in CI, probably need to use **Dockerfile.upstream.c9s**.

Koncpa commented on Library/test-helpers/Dockerfile.registrar:
```diff
@@ -1,9 +1,12 @@
 FROM quay.io/centos/centos:stream9
 COPY id_*.pub /root/
-COPY lime_con_start.sh /usr/local/bin/lime_con_start
```
When you remove copying script for starting several processes/services like (sshd and keylime) the additional setup of SSH is useless.

Koncpa commented on Library/test-helpers/Dockerfile.upstream.c9s:
```diff
@@ -1,5 +1,4 @@
 FROM quay.io/centos/centos:stream9
-COPY lime_con_start.sh /usr/local/bin/lime_con_start
```
Same as in previous docker file.

Koncpa commented on Library/test-helpers/Dockerfile.verifier:
```diff
@@ -1,9 +1,11 @@
 FROM quay.io/centos/centos:stream9
 COPY id_*.pub /root/
-COPY lime_con_start.sh /usr/local/bin/lime_con_start
```
Same as in previous docker file.

kkaarreell reviewed (changes requested):
I think there are few details that we should probably discuss.

ansasaki commented on Library/test-helpers/Dockerfile.upstream.c9s:
```diff
@@ -1,11 +1,11 @@
+EXPOSE 9002
+EXPOSE 8892
+
+CMD ["keylime_agent"]
```
I added this because I thought this image was used only for the agent. Probably I'll propose completely removing this Dockerfile later, and use the actual upstream Dockerfiles instead (those maintained as part of the upstream repo).

kkaarreell replied:
Can we use upstream Dockerfiles already or are there any changes still needed?
Also, in order to work with upstream Dockerfiles, maybe we could support storing Dockerfile URL in the `VERIFIER_DOCKERFILE` variable.

ansasaki replied:
I'm still fighting a permission issue: I need the keylime user inside the container to have access to the certificate files generated outside the container. 

The problem is that the keylime user inside the container has an arbitrary UID, and I have to change the ownership accordingly. 

I need a way to get the UID of the keylime user inside the container so that I `chown` the files before mounting, or a way to map the keylime user outside the container to the keylime user inside the container.

The only solution I have now is ugly:

- Run the `id -u keylime` command inside the container once to get the UID, `chown` the files, and then start the container.
- Alternatively, instead of `chown` the files to the UID of the keylime user inside the container, I can use `--uidmap` option to map the keylime user outside to the keylime user inside, but I still need the UID of the keylime user inside the container.

It would be great if I could use something like `podman unshare` to do this, but this only exist when running containers as non-root (and we need to run as root to have access to the TPM and event logs). Even better if I could map the UIDs using the username inside the container (something like `--usermap=<host username>:<container username>`).

I believe this is something that affects many services that are executed as root, but later drops privileges to an unprivileged user. Do you have any suggestion?

Koncpa replied:
Hmm I don't know if it's possible, but what about run container as non-root and add root capabilities for access to TPM and event log?

ansasaki replied:
> Hmm I don't know if it's possible, but what about run container as non-root and add root capabilities for access to TPM and event log?

This is the dream, to be able to run with unprivileged user and still have access to TPM and event log. If possible to do this by adding capabilities, then our work would be much easier.

Koncpa replied:
I'll contact the podman guy and ask him about this possibility and let you know.

ansasaki replied:
Just for the record: we were informed that capabilites in non-root container are just namespaced capabilities, not real capabilities. This makes it not possible to run the agent in the container as an unprivileged user and provide capabilities to access resources that require higher privileges.

ansasaki commented on Library/test-helpers/lib.sh:
```diff
@@ -2062,15 +2052,10 @@ limeconRun() {
     local TAG=$2
     local IP=$3
     local NETWORK=$4
-    local COMMAND=$5
```
I think we can put this back for now, but the upstream (and production images), this should not be needed, meaning that the container has a single purpose and well defined entry point

ansasaki commented on Library/test-helpers/lib.sh:
```diff
@@ -2163,8 +2199,32 @@ limeconRunRegistrar() {
+        EXTRA_ARGS="--volume $CERTDIR:/var/lib/keylime/cv_ca:z $EXTRA_ARGS"
+    fi
+
+    limeconRun $NAME $TAG $IP $NETWORK "$EXTRA_ARGS"
```
As in other comments, I think we should use single purpose images as the production images will be.

ansasaki commented on container/functional/keylime_ipv6_multihost/test.sh:
```diff
@@ -7,6 +7,10 @@ HTTP_SERVER_PORT=8080
 # set REVOCATION_NOTIFIER=zeromq to use the zeromq notifier
 [ -n "$REVOCATION_NOTIFIER" ] || REVOCATION_NOTIFIER=agent
 
+VERIFIER_DOCKERFILE="Dockerfile.verifier"
```
For English readers, I think the `XXX_DOCKERFILE` format is easier to understand. I propose changing this everywhere.

kkaarreell replied:
Ack.

ansasaki commented on container/functional/keylime_verifier_registrar_container-basic-attestation/test.sh:
```diff
@@ -7,9 +7,9 @@
-[ -n "$DOCKERFILE_VERIFIER" ] || DOCKERFILE_VERIFIER=Dockerfile.upstream.c9s
-[ -n "$DOCKERFILE_REGISTRAR" ] || DOCKERFILE_REGISTRAR=Dockerfile.upstream.c9s
-[ -n "$DOCKERFILE_AGENT" ] || DOCKERFILE_AGENT=Dockerfile.upstream.c9s
+[ -n "$VERIFIER_DOCKERFILE" ] || VERIFIER_DOCKERFILE=Dockerfile.verifier
```
Again, I think we should change the variables to be in `XXX_DOCKERFILE` format everywhere.

ansasaki commented:
@kkaarreell @Koncpa I think this is ready for review. I removed most of the customization done to the images. The only remaining workaround left was for the issue with the permissions: the agent container is executed by root, and will drop privileges to the internal `keylime` user. The problem is that the internal user needs access to the CA certificates. The workaround I added is a podman run that `chown` the files in the volume mounted in `/var/lib/keylime/cv_ca`.

This workaround is applied to the built image, not on the Dockerfile. Theoretically, we could test the upstream Dockerfiles instead of the ones we created.

I'll start working on the other part, where we fetch built images from a registry instead of building from the Dockerfile.

Koncpa commented:
@ansasaki I'm not sure, if won't be needed provide additional changes in [keylime server role test repo](https://github.com/RedHat-SP-Security/keylime_server-role-tests) and also rework your PR. There is used ssh access to container  for conecting ansible for preparing keylime roles. It seems that in your PR try to remove all container customization related to ssh. Please correct me if I'm wrong.
FYI:  @kkaarreell

Koncpa commented on Library/test-helpers/lib.sh:
```diff
@@ -2006,12 +2005,7 @@ limeconPrepareImage() {
         ARGS="--volume /var/tmp/keylime_sources:/mnt/keylime_sources:z"
     fi
 
-    #set up for ssh access
```
Probably that should be keep it there regarding to containers for keylime ansible roles testing or rework it.

ansasaki commented:
> @ansasaki I'm not sure, if won't be needed provide additional changes in [keylime server role test repo](https://github.com/RedHat-SP-Security/keylime_server-role-tests) and also rework your PR. There is used ssh access to container for conecting ansible for preparing keylime roles. It seems that in your PR try to remove all container customization related to ssh. Please correct me if I'm wrong. FYI: @kkaarreell

Yes, I'm trying to remove all customization, including the SSH setup from the test image, because the keylime images won't have that. This way we can use the same tests for the images we build as part of the test (from the dockerfiles) as well as for the images we obtain directly from the registry.

If we rely on special setup (e.g. to add SSH access), then it would be difficult to reuse the tests for the images obtained from the registry.

If the SSH access is strictly necessary, then we could create an image by adding the SSH setup on a layer on top of the tested image (used as a base image) for the test cases that need SSH. It is not ideal, but at least we could use the keylime images obtained from the registry as well as the test images built as part of the tests.

Koncpa commented:
Understand..Hm for rest of the test images is not necessary to have ssh access, just for systemd image which is use for keylime ansible role tests. Systemd image should have all necessary ssh setup in docker file, but we need somewhere else generate ssh keys which are essential for access and for now it's directly generate in func limeconPrepareImage. There are two options: create new function in lib for generating ssh keys or in every ansible role test generate ssh keys. Probably it would be better to have function for generating ssh keys...WDYT?? @kkaarreell

kkaarreell commented:
I think this is fine. As you said, ordinary images do not need ssh access and for the systemd container (that we use for system roles testing) the ssh setup remains in place. So I believe nothing should be missing. We could possibly run system roles tests against Anderson's branch to double check.

kkaarreell commented:
> We could possibly run system roles tests against Anderson's branch to double check.
I just realized it won't be possible until we rebase keylime in RHEL since here we are targeting upstream while keylime_server role is for RHEL, i.e. the test code is currently not compatible.

Koncpa commented:
> I think this is fine. As you said, ordinary images do not need ssh access and for the systemd container (that we use for system roles testing) the ssh setup remains in place. So I believe nothing should be missing. We could possibly run system roles tests against Anderson's branch to double check.


@kkaarreell  But [part](https://github.com/RedHat-SP-Security/keylime-tests/blob/f8cd27f7821b710f1038010030d3e7d493a9272a/Library/test-helpers/lib.sh#L2010) of the ssh setup on host is removed in this PR, so we neeed to update it.

ansasaki commented:
> > I think this is fine. As you said, ordinary images do not need ssh access and for the systemd container (that we use for system roles testing) the ssh setup remains in place. So I believe nothing should be missing. We could possibly run system roles tests against Anderson's branch to double check.
> 
> @kkaarreell But [part](https://github.com/RedHat-SP-Security/keylime-tests/blob/f8cd27f7821b710f1038010030d3e7d493a9272a/Library/test-helpers/lib.sh#L2010) of the ssh setup on host is removed in this PR, so we neeed to update it.

I can put back the key generation part, but I think it would be better if either it is conditional, controlled by some environment variable, or in a separate helper function. Otherwise new keys would be generated for each built image, even though they are not used in most of them.

kkaarreell commented:
I think that piece can be put into the respective test.

Koncpa commented:
[FYI: Add ssh keygen to keylime server roles test repo.](https://github.com/RedHat-SP-Security/keylime_server-role-tests/pull/7)

kkaarreell commented:
Hi @ansasaki , do you plan further updates or can we merge it?

ansasaki commented:
> Hi @ansasaki , do you plan further updates or can we merge it?

I think we can merge this. I'll open a separate PR for next changes
